signwriting_animation/scripts/minimal_loop.py

```diff
@@ -0,0 +1,265 @@
+    x = _to_dense(x)
+    if x.dim() == 5:  # [B,T,P,J,C]
+        x = x[:, :, 0, ...]
+    if x.dim() != 4:
+        raise ValueError(f"Expected 4D tensor [B,T,J,C], got {tuple(x.shape)}")
+    return x
+
+
+def btjc_to_bjct(x):
+    """Convert [B, T, J, C] → [B, J, C, T] (model forward format)."""
+    return x.permute(0, 2, 3, 1).contiguous()
+
+
+def bjct_to_btjc(x):
+    """Convert [B, J, C, T] → [B, T, J, C] (loss/metrics format)."""
+    return x.permute(0, 3, 1, 2).contiguous()
+
+
+def masked_mse(pred, tgt, mask_bt):
+    pred, tgt = sanitize_btjc(pred), sanitize_btjc(tgt)
+
+    Tm = min(pred.size(1), tgt.size(1), mask_bt.size(1))
+    pred = pred[:, :Tm]
+    tgt  = tgt[:,  :Tm]
+
+    m4 = mask_bt[:, :Tm].float()[:, :, None, None]   # [B,T,1,1]
+
+    diff2 = (pred - tgt) ** 2                        # [B,T,J,C]
+    num = (diff2 * m4).sum()
+    den = (m4.sum() * pred.size(2) * pred.size(3)).clamp_min(1.0)
+    return num / den
+
+def masked_dtw(pred_btjc, tgt_btjc, mask_bt):
+    pred = sanitize_btjc(pred_btjc)  # [B,T,J,C]
+    tgt  = sanitize_btjc(tgt_btjc)
+    B, T, J, C = pred.shape
+    vals = []
+
+    for b in range(B):
+        t = min(int(mask_bt[b].sum().item()), pred.size(1), tgt.size(1))
+        if t <= 1:
+            continue
+
+        x = pred[b, :t].reshape(t, -1).to(dtype=torch.float32)  # [t, J*C]
+        y = tgt[b,  :t].reshape(t, -1).to(dtype=torch.float32)
+
+        device = x.device
+        D = torch.cdist(x, y)                      # [t, t]
+        Cmat = torch.empty((t, t), device=device)
+        Cmat[0, 0] = D[0, 0]
+        for i in range(1, t):
+            Cmat[i, 0] = D[i, 0] + Cmat[i-1, 0]
+        for j in range(1, t):
+            Cmat[0, j] = D[0, j] + Cmat[0, j-1]
+
+        for i in range(1, t):
+            for j in range(1, t):
+                m = torch.minimum(
+                        torch.minimum(Cmat[i-1, j], Cmat[i, j-1]),
+                        Cmat[i-1, j-1]
+                    )
+                Cmat[i, j] = D[i, j] + m
+
+        vals.append(Cmat[t-1, t-1] / (2 * t))
+
+    if not vals:
+        return torch.tensor(0.0, device=pred.device)
+    return torch.stack(vals).mean()
+
+
+class FilteredDataset(Dataset):
+    """Only keep valid samples."""
+    def __init__(self, base: Dataset, target_count=200, max_scan=5000):
+        self.base = base
+        self.idx = []
+        N = len(base)
+        for i in range(min(N, max_scan)):
+            try:
+                it = base[i]
+                if isinstance(it, dict) and "data" in it and "conditions" in it:
+                    self.idx.append(i)
+                if len(self.idx) >= target_count:
+                    break
+            except Exception:
+                continue
+        if not self.idx:
+            self.idx = [0]
+
+    def __len__(self):
+        return len(self.idx)
+
+    def __getitem__(self, i):
+        return self.base[self.idx[i]]
+
+class LitMinimal(pl.LightningModule):
```

Comment:
generally, this class should not be in "scripts", it can be `lightning_module.py` or something.